### PR TITLE
NME for Gaussian Splatting

### PR DESCRIPTION
**Pull request by CedricGuillemet**
This snippet: #CSD1ZK#1
will show this:
https://github.com/user-attachments/assets/65fa9f69-2eb1-4115-8cc8-a80cb5969f2e

### CODE REVIEW

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import { NodeMaterialBlock } from "../../nodeMaterialBlock";
+import { NodeMaterialBlockConnectionPointTypes } from "../../Enums/nodeMaterialBlockConnectionPointTypes";
+import type { NodeMaterialBuildState } from "../../nodeMaterialBuildState";
+import { NodeMaterialBlockTargets } from "../../Enums/nodeMaterialBlockTargets";
+import type { NodeMaterialConnectionPoint } from "../../nodeMaterialBlockConnectionPoint";
+import { RegisterClass } from "../../../../Misc/typeStore";
+
+/**
+ * Block used for the particle ramp gradient section
```
Comment (deltakosh) [inline packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts:9]: wrong comment

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * Gets the scale input component
+     */
+    public get scale(): NodeMaterialConnectionPoint {
+        return this._inputs[1];
+    }
+    /**
+     * Gets the color input component
+     */
+    public get color(): NodeMaterialConnectionPoint {
```
Comment (deltakosh) [inline packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts:51]: should be removed?

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        return this._inputs[2];
+    }
+
+    /**
+     * Gets the splatColor output component
```
Comment (deltakosh) [inline packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts:56]: splat color or splat vertex?

**Review by deltakosh — COMMENTED** [review]
Some copy paste mistakes:)

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15677/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15677/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15677/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15677/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15677/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15677/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: .vscode/launch.json
Comment (RaananW) [inline .vscode/launch.json:1]: please revert the changes here :-)

Comment (CedricGuillemet) [reply .vscode/launch.json:1]: oops!

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15677/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianBlock.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            return;
+        }
+
+        // Emit code
+        const comments = `//${this.name}`;
+        state._emitFunctionFromInclude("clipPlaneFragmentDeclaration", comments);
+        state._emitFunctionFromInclude("logDepthDeclaration", comments);
+        state._emitFunctionFromInclude("fogFragmentDeclaration", comments);
+        state._emitFunctionFromInclude("gaussianSplattingFragmentDeclaration", comments);
+        state._emitVaryingFromString("vPosition", NodeMaterialBlockConnectionPointTypes.Vector2);
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianBlock.ts:69]: "vPosition" should be added to the list of excluded names (`state._excludeVariableName`).

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/splatReaderBlock.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        state._emitVaryingFromString("vPosition", NodeMaterialBlockConnectionPointTypes.Vector2);
+        state._emitUniformFromString("dataTextureSize", NodeMaterialBlockConnectionPointTypes.Vector2);
+        const splatIndex = this.splatIndex;
+        const splatPosition = this.splatPosition;
+        const splatColor = this.splatColor;
+
+        const splatVariablename = state._getFreeVariableName("splat");
+        state.compilationString += `Splat ${splatVariablename} = readSplat(${splatIndex.associatedVariableName});\n`;
+
+        state.compilationString += "vec3 covA = splat.covA; vec3 covB = splat.covB;\n";
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/splatReaderBlock.ts:91]: "covA" and "covB" should also be added to `_excludeVariableName`.

File: packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts [hunk 6]
```diff
@@ -75,6 +79,39 @@ export class GaussianSplattingMesh extends Mesh {
...
+    /**
+     * set rendering material
+     */
+    public override set material(value: Material) {
+        this._material = value;
+        this._material.backFaceCulling = true;
+        this._material.cullBackFaces = false;
+        this._clearObservable();
+        if (value.getClassName() === "NodeMaterial") {
+            value.onBindObservable.add(this._materialBindCallback, undefined, undefined, this);
```
Comment (Popov72) [inline packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts:91]: I think it would be cleaner to do the bindings this way:

https://github.com/Popov72/Babylon.js/commit/916b08e80725ff49afd03e5a01a0bae1ce459e40

**Review by Popov72 — CHANGES_REQUESTED** [review]

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/splatReaderBlock.ts (hunk 5 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/splatReaderBlock.ts:91]: already added

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianBlock.ts (hunk 4 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianBlock.ts:69]: Fixed

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+        if (state.target === NodeMaterialBlockTargets.Fragment) {
+            return;
+        }
+
+        const comments = `//${this.name}`;
+        state._emitFunctionFromInclude("gaussianSplattingVertexDeclaration", comments);
+        state._emitFunctionFromInclude("gaussianSplatting", comments);
+        state._emitUniformFromString("focal", NodeMaterialBlockConnectionPointTypes.Vector2);
+        state._emitUniformFromString("invViewport", NodeMaterialBlockConnectionPointTypes.Vector2);
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianSplattingBlock.ts:98]: Fixed

File: packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianBlock.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        return this._outputs[0];
+    }
+
+    /**
+     * Initialize the block and prepare the context for build
+     * @param state defines the state that will be used for the build
+     */
+    public override initialize(state: NodeMaterialBuildState) {
+        //state._excludeVariableName("splatColor");
+    }
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Materials/Node/Blocks/GaussianSplatting/gaussianBlock.ts:54]: Fixed

File: packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts (hunk 6 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts]: It looks better, indeed!

File: packages/dev/core/src/Materials/GaussianSplatting/gaussianSplattingMaterial.ts [hunk 9]
```diff
@@ -177,6 +177,57 @@ export class GaussianSplattingMaterial extends PushMaterial {
         return true;
     }
 
+    /**
+     *
+     * @param mesh
+     * @param effect
+     * @param scene
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Materials/GaussianSplatting/gaussianSplattingMaterial.ts]: fixed

**Review by Popov72 — APPROVED** [review]
LGTM!

Have you been able to test in WebGPU?

**Comment by CedricGuillemet** [comment]
> LGTM!
> 
> Have you been able to test in WebGPU?

PG is fine with standard GS and nodematerial but in NME, the preview never finishes loading the GS; even though I can see it  being loaded. The message "Please wait" doesn't disappear. I'm taking a look.

File: packages/dev/core/src/Materials/Node/nodeMaterial.ts [hunk 10]
```diff
@@ -2153,6 +2156,65 @@ export class NodeMaterial extends PushMaterial {
         this._mode = NodeMaterialModes.Particle;
     }
 
+    /**
+     * Clear the current material and set it to a default state for gaussian splatting
+     */
+    public setToDefaultGaussianSplatting() {
```
Comment (deltakosh) [inline packages/dev/core/src/Materials/Node/nodeMaterial.ts:2162]: My only ask here would be to make that an external function so we do not grow the size of nodeMaterial whennot using GS. I should have done that for other modes but it is too late now

**Review by deltakosh — APPROVED** [review]

File: packages/dev/core/src/Materials/Node/nodeMaterial.ts (hunk 10 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Materials/Node/nodeMaterial.ts]: I've moved the function into `nodeMaterialDefault.ts`

Comment (CedricGuillemet) [reply packages/dev/core/src/Materials/Node/nodeMaterial.ts]: Merging now.
